msteams-platform/tabs/how-to/dashboard-teams-toolkit.md

```diff
@@ -0,0 +1,991 @@
+            const provider = new TeamsFxProvider(TeamsUserCredentialContext.getInstance().getCredential(), [
+              "Tasks.ReadWrite",
+            ]);
+            Providers.globalProvider = provider;
+        
+            // Check if user is signed in
+            if (await this.checkIsConsentNeeded()) {
+              await loginAction(["Tasks.ReadWrite"]);
+            }
+        
+            // Update signed in state
+            Providers.globalProvider.setState(ProviderState.SignedIn);
+            this.setState({ needLogin: false });
+          }
+        
+          /**
+           * Check if user needs to consent
+           * @returns true if user needs to consent
+           */
+          async checkIsConsentNeeded() {
+            let needConsent = false;
+            try {
+              await TeamsUserCredentialContext.getInstance().getCredential().getToken(["Tasks.ReadWrite"]);
+            } catch (error) {
+              needConsent = true;
+            }
+            return needConsent;
+          }
+        }
+        
+      ```
+
+      For more information, refer [Microsoft Graph Toolkit](/graph/toolkit/overview).
+
+1. Add the widget to dashboard layout. Include the new widget in your dashboard file:
+
+```typescript
+...
+export default class YourDashboard extends Dashboard {
+  ...
+  protected dashboardLayout(): undefined | JSX.Element {
+    return (
+      <>
+        <GraphWiget />
+      </>
+    );
+  }
+  ...
+}
+```
+
+Now, launch or refresh your Teams app, you'll see the new widget using Microsoft Graph Toolkit.
+
+### How to include a data loader
+
+If you want to include a data loader to your widget before the widget is loaded, you can add a property to the state of the widget to indicate that the data loader is loading. This property can be used to show a loading indicator to the user.
+
+The following steps show how to add a property to the state of `ListWidget` and how to use it to show a loading spinner while the data is loading.
+
+1. **Define a state type**
+
+   Define a state type including a property named `loading` that indicates whether the data is loading.
+
+    ```typescript
+    interface ListWidgetState {
+      data: ListModel[];
+      loading?: boolean;
+    }
+    ```
+
+1. **Add a data loader**
+
+    Modify the `bodyContent` method to show a loading spinner if data is loading.
+
+    ```typescript
+    bodyContent(): JSX.Element | undefined {
+      return (
+        <>
+          {this.state.loading !== false ? (
+            <div style={{ display: "grid", justifyContent: "center", height: "100%" }}>
+              <Spinner label="Loading..." labelPosition="below" />
+            </div>
+          ) : (
+            <div style={bodyContentStyle()}>
+              ...
+            </div>
+          )}
+        </>
+      );
+    }
+    ```
+
+1. **Hide the footer button if the data is loading**
+
+    Following is an example of footer button:
+
+    ```typescript
```

Comment:
It may be better if the language of this code snippet is adjusted to `tsx`.

Reply:
Updated to `tsx`